From db375e56763a17bfaf997f4ed1bdcfcdb3458154 Mon Sep 17 00:00:00 2001
From: Narendhar Reddy <quic_nare@quicinc.com>
Date: Sun, 24 Nov 2024 11:00:28 +0530
Subject: [PATCH] QSDK:PROTCOL mac80211: Reorder work flushes after
 unlocking wiphy and rtnl locks

Reorder the flush operations for `awgn_detected_work` and `cw_detected_work`
in `ieee80211_unregister_hw` to occur after unlocking the wiphy and rtnl locks.
This change ensures that the work flushes are performed without holding
these locks, potentially preventing deadlocks and improving the overall
stability of the unregister process.

Signed-off-by: Narendhar Reddy <quic_nare@quicinc.com>
---
 net/mac80211/main.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/mac80211/main.c b/net/mac80211/main.c
index 64f3062..3ec6012 100644
--- a/net/mac80211/main.c
+++ b/net/mac80211/main.c
@@ -1620,11 +1620,11 @@ void ieee80211_unregister_hw(struct ieee80211_hw *hw)
 	wiphy_delayed_work_cancel(local->hw.wiphy, &local->roc_work);
 	wiphy_work_cancel(local->hw.wiphy, &local->sched_scan_stopped_work);
 	wiphy_work_cancel(local->hw.wiphy, &local->radar_detected_work);
-	flush_work(&local->awgn_detected_work);
-	flush_work(&local->cw_detected_work);
 	wiphy_unlock(local->hw.wiphy);
 	rtnl_unlock();
 
+	flush_work(&local->awgn_detected_work);
+	flush_work(&local->cw_detected_work);
 	cancel_work_sync(&local->restart_work);
 	cancel_work_sync(&local->reconfig_filter);
 
-- 
2.34.1

